docs/roadmap/scaling-2019.md

```diff
@@ -0,0 +1,115 @@
+            env:
+            - name: SLEEP_TIME
+              value: "100ms"
+```
+```bash
+for i in `seq 1 1000`; do curl http://$MY_IP & done
+```
+
+This verifies that 1) we can handle all requests in an overload without error and 2) all the requests don't land on a single pod, which would take 100 sec.
+
+* POC: Vadim Raskin (IBM)
+* Github: [Project 7](https://github.com/knative/serving/projects/7)
+
+### Autoscaler Availability
+
+Because Knative scales to zero, autoscaling is in the critical-path for serving requests. If the autoscaler isn't available when an idle Revision receives a request, that request will not be served. Other components such as the Activator are in this situation too. But they are more stateless and so can be scaled horizontally relatively easily. For example, any Activator can proxy any request for any Revision. All it has to do is send a messge to the Autoscaler and then wait for a Pod to show up. Then it proxies the request and is taken back out of the serving path.
+
+However the Autoscaler process is more stateful. It maintains request statistics over a window of time. And it must process data from the Revision Pods continuously to maintain that window of data. It is part of the running system all the time, not just when scaled to zero. As the number of Revisions increase and the number of Pods in each Revision increases, the CPU and memory requirements will exceed that available to a single process. So some sharding is necessary.
+
+**Our goal is to shard Revisions across Autoscaler replicas (e.g. 2x the Replica count means 1/2 the load on each Autoscaler). And for autoscaling to be unaffected by an individual Autoscaler termination.**
+
+* POC: Kenny Leung (Google)
+* Github: [Project 19](https://github.com/knative/serving/projects/19)
+
+### Streaming Autoscaling
+
+In addition to being always available, Web applications are expected to be responsive. Long-lived connections and streaming protocols like Websockets and HTTP2 are essential. Knative Serving accepts HTTP2 connections. And will serve requests multiplexed within the connection. But the autoscaling subsystem doesn't quite know what to do with those connections. It sees each connection as continuous load on the system and so will autoscale accordingly.
+
+But the actual load is in the stream within the connection. So the metrics reported to the Autoscaler should be based on the number of concurrent *streams*. This requires some work in the Queue proxy to crack open the connection and emit stream metrics.
+
+Additionally, concurrency limits should be applied to streams, not connections. So containers which can handle only one request at at time should still be able to serve HTTP2. The Queue proxy will just allow one stream through at a time.
+
+**Our goal is 1) to support HTTP2 end-to-end while scaling on concurrent streams and in this mode 2) enforce concurrency limits on streams (not connections).**
```

Comment:
I'd like @evankanderson to review this second in particular, since his knowledge of http/2 is >>>>>> my own.

Reply:
Yes, input please!

Reply:
If @evankanderson could join the Autoscaling WG meeting today, that'd be awesome. I've got some bits to share there.

Reply:
I've dropped this line item from the roadmap.